src/rt/GEAR/rt_grackle_cooling.h

```diff
@@ -0,0 +1,228 @@
+ * @param rho particle physical density
+ * @param particle_grackle_data The grackle_field_data structure from grackle.
+ **/
+__attribute__((always_inline)) INLINE static void
+rt_tchem_copy_from_grackle1(struct part* p, gr_float rho,
+                               grackle_field_data* data) {
+  
+  const gr_float one_over_rho = 1. / rho;
+
+  p->rt_data.tchem.mass_fraction[rt_species_HI] =
+      *data->HI_density * one_over_rho;
+
+  p->rt_data.tchem.mass_fraction[rt_species_HII] =
+      *data->HII_density * one_over_rho;
+
+  p->rt_data.tchem.mass_fraction[rt_species_HeI] =
+      *data->HeI_density * one_over_rho;
+  
+  p->rt_data.tchem.mass_fraction[rt_species_HeII] =
+      *data->HeII_density * one_over_rho;
+  
+  p->rt_data.tchem.mass_fraction[rt_species_HeIII] =
+      *data->HeIII_density * one_over_rho;
+  
+  p->rt_data.tchem.mass_fraction[rt_species_e] =
+      *data->e_density * one_over_rho;
+
+}
+
+
+/**
+ * @brief copy the grackle data to particle data for grackle chemistry mode 2.
+ *
+ * @param p particle to use
+ * @param rho particle physical density
+ * @param particle_grackle_data The grackle_field_data structure from grackle.
+ **/
+__attribute__((always_inline)) INLINE static void
+rt_tchem_copy_from_grackle2(struct part* p, gr_float rho,
+                               grackle_field_data* data) {
+ 
+  const gr_float one_over_rho = 1. / rho;
+  double rhodust;
+
+  p->rt_data.tchem.mass_fraction[rt_species_HM] =
+      *data->HM_density * one_over_rho;
+
+  p->rt_data.tchem.mass_fraction[rt_species_H2I] =
+      *data->H2I_density * one_over_rho;
+
+  p->rt_data.tchem.mass_fraction[rt_species_H2II] =
+      *data->H2II_density * one_over_rho;
+
+#ifdef SWIFT_RT_GRACKLE_DUST
+  /* Load gas metallicities */
+      p->chemistry_data.metal_mass_fraction[1] = *data->He_gas_metalDensity * one_over_rho;
+      p->chemistry_data.metal_mass_fraction[2] = *data->C_gas_metalDensity * one_over_rho;
+      p->chemistry_data.metal_mass_fraction[3] = *data->N_gas_metalDensity * one_over_rho;
+      p->chemistry_data.metal_mass_fraction[4] = *data->O_gas_metalDensity * one_over_rho;
+      p->chemistry_data.metal_mass_fraction[5] = *data->Ne_gas_metalDensity * one_over_rho;
+      p->chemistry_data.metal_mass_fraction[6] = *data->Mg_gas_metalDensity * one_over_rho;
+      p->chemistry_data.metal_mass_fraction[7] = *data->Si_gas_metalDensity * one_over_rho;
+      p->chemistry_data.metal_mass_fraction[8] = *data->S_gas_metalDensity * one_over_rho;
+      p->chemistry_data.metal_mass_fraction[9] = *data->Ca_gas_metalDensity * one_over_rho;
+      p->chemistry_data.metal_mass_fraction[10] = *data->Fe_gas_metalDensity * one_over_rho;
+      /* Load dust metallicities */
+      p->rt_data.cooling.dust_mass = *data->dust_density * p->mass * one_over_rho;
+      p->rt_data.cooling.dust_mass_fraction[0] = 0.f;
+      for (int i=1; i<chemistry_element_count; i++) {
+	  if (i==1) rhodust = *data->He_dust_metalDensity;
```

Comment:
There is no point in doing a for loop if you have to manually select something each time...

Reply:
I just copied this part from kiara cooling. I can optimise it later after finishing all